ext_proc: Add code to check mutation rules

gbrail opened the pull request:
Commit Message: Add a library to check against the rules in the new HeaderMutationRules proto.
Risk Level: Low -- code is not called anywhere yet.
Testing: New unit tests.

Signed-off-by: Gregory Brail <gregbrail@google.com>

gbrail commented:
This PR will be used in ext_proc, but I put it in a "common" location so that we could potentially use it elsewhere along with the new proto. Could either @snowp or @htuch take a look? Thanks!

htuch commented on source/extensions/filters/common/mutation_rules/mutation_rules.cc:
```diff
@@ -0,0 +1,89 @@
+  return routing_headers;
+}
+
+const absl::flat_hash_set<Http::LowerCaseString>& Checker::getRoutingHeaders() {
```
This can be done with the `CONSTRUCT_ON_FIRST_USE` macro.

gbrail replied:
Thanks! Used the macro and followed the pattern.

htuch reviewed (commented):
LGTM modulo nit.

htuch commented on source/extensions/filters/common/mutation_rules/mutation_rules.cc:
```diff
@@ -0,0 +1,97 @@
+
+private:
+  absl::flat_hash_set<LowerCaseString> headers_;
+};
```
This could just be a function returning an `absl::flat_hash_set`, but this works as well..

htuch reviewed (commented):
LGTM, thanks. @snowp can you take a quick look as a non-Googler? Thanks.

gbrail commented:
FWIW, the CI failure is in kill_request:kill_request_filter_integration_test, and doesn't have anything to do with this PR.

gbrail commented:
Update -- I added a few more rules to make it easier to replace all the logic that is today inside the ext_proc filter itself. I also changed the docs in the .proto a bit which is what triggered the API review.

snowp reviewed (approved):
LGTM, thanks!
